lib/dyno.js

```diff
@@ -62,28 +66,102 @@ class Dyno {
    */
   attach () {
     return new Promise((resolve, reject) => {
-      if (this.opts.showStatus) cli.action.status(this._status('starting'))
       this.resolve = resolve
       this.reject = reject
-      let uri = url.parse(this.dyno.attach_url)
-      let c = tls.connect(uri.port, uri.hostname, {rejectUnauthorized: this.heroku.options.rejectUnauthorized})
-      c.setTimeout(1000 * 60 * 20)
-      c.setEncoding('utf8')
-      c.on('connect', () => {
-        c.write(uri.path.substr(1) + '\r\n', () => {
-          if (this.opts.showStatus) cli.action.status(this._status('connecting'))
+      this.uri = url.parse(this.dyno.attach_url)
+
+      if (this.uri.protocol === 'http:' || this.uri.protocol === 'https:') {
+        this._ssh()
+      } else {
+        this._rendezvous()
+      }
+    })
+  }
+
+  _rendezvous () {
+    if (this.opts.showStatus) cli.action.status(this._status('starting'))
+    let c = tls.connect(this.uri.port, this.uri.hostname, {rejectUnauthorized: this.heroku.options.rejectUnauthorized})
+    c.setTimeout(1000 * 60 * 20)
+    c.setEncoding('utf8')
+    c.on('connect', () => {
+      c.write(this.uri.path.substr(1) + '\r\n', () => {
+        if (this.opts.showStatus) cli.action.status(this._status('connecting'))
+      })
+    })
+    c.on('data', this._readData(c))
+    c.on('close', () => {
+      this.opts['exit-code'] ? this.reject('No exit code returned') : this.resolve()
+      if (this.unpipeStdin) this.unpipeStdin()
+    })
+    c.on('error', this.reject)
+    process.once('SIGINT', () => c.end())
+  }
+
+  _ssh () {
+    this._wait()
```

Comment:
I've since revised things. it works now. i'll send you some instructions to try out the changes